### ISSUE TITLE
Extending value inlining

### ISSUE THREAD

**Issue by guybedford** [rollup/rollup#2152]
From discussions it sounds like we're much closer to enabling some pretty amazing value inlining optimizations than I thought we were.

We did have some previous discussion here at https://github.com/rollup/rollup/issues/1904, but it is worth going over some of these points again.

One of the things to stress here again is that what we care most about optimizing is the *top-level* code execution critical path. Tree shaking already handles the other cases pretty well so tracking analysis can be specifically tailored to understand this path first and foremost I think.

I just wanted to list some sample scenarios here, and then hopefully get some feedback on:

a) How hard these cases would be to support given the current architecture
b) How useful these cases would be - as we do have to balance performance here somewhat.

The main use cases are rather broad, I expect we will discuss the intricacies of each one in depth, but just trying to get the lay of the land here first in terms of priorities.

I've probably already got ahead of things with (5) and (6), but if we were to really cast out this net, what else might I have missed here?

1. Path value inlining - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHglMjBmcm9tJTIwJy4lMkZqc29uJyUzQiU1Q25jb25zb2xlLmxvZyh4LnkueiklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyanNvbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBkZWZhdWx0JTIwJTdCJTVDbiU1Q3R5JTNBJTIwJTdCJTVDbiU1Q3QlNUN0eiUzQSUyMDUlNUNuJTVDdCU3RCUyQyU1Q24lNUN0ciUzQSUyMCU3QiU1Q24lNUN0JTVDdHRyZWVzaGFrZSUzQSUyMCdtZSclNUNuJTVDdCU3RCU1Q24lN0QlM0IlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

2. Known value inlining in simple expressions - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMHglMjBmcm9tJTIwJy4lMkZqc29uJyUzQiU1Q25pZiUyMCh4JTIwJTI2JTI2JTIweCUyMCElM0QlM0QlMjB0cnVlKSU1Q24lNUN0Y29uc29sZS5sb2coJ3RyZWVzaGFrZSUyMG1lJyklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyanNvbi5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBkZWZhdWx0JTIwdHJ1ZSUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA== 

Ideally an interface for applying known global constant values like `process.env.NODE_ENV = 'production'` setting for example.

We should probably try to fully flesh out and categorize some of the basic cases here much further as it seems like this is the primary point where we are at right now.

3. Recursive value inlining with treeshaking - once we go down this road it gets its own depths as to how deep to apply, and knowing if each optimization opens up further ones - so perhaps having passes like the treeshaking will make sense? I found a good example of something we could improve in the React codebase itself. The build file contains some redundant code output - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMid1c2UlMjBzdHJpY3QnJTNCJTVDbiU1Q25pbXBvcnQlMjBSZWFjdCUyMGZyb20lMjAnLiUyRnNyYyUyRlJlYWN0JyUzQiU1Q24lNUNuJTJGJTJGJTIwVE9ETyUzQSUyMGRlY2lkZSUyMG9uJTIwdGhlJTIwdG9wLWxldmVsJTIwZXhwb3J0JTIwZm9ybS4lNUNuJTJGJTJGJTIwVGhpcyUyMGlzJTIwaGFja3klMjBidXQlMjBtYWtlcyUyMGl0JTIwd29yayUyMHdpdGglMjBib3RoJTIwUm9sbHVwJTIwYW5kJTIwSmVzdC4lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMFJlYWN0LmRlZmF1bHQlMjAlM0YlMjBSZWFjdC5kZWZhdWx0JTIwJTNBJTIwUmVhY3QlM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyc3JjJTJGUmVhY3QuanMlMjIlMkMlMjJjb2RlJTIyJTNBJTIyY29uc3QlMjBSZWFjdCUyMCUzRCUyMCU3QiU1Q24lMjAlMjBDb21wb25lbnQlM0ElMjAlN0IlN0QlMkMlNUNuJTIwJTIwdmVyc2lvbiUzQSUyMCcxLjIuMyclMkMlNUNuJTIwJTIwX19TRUNSRVRfSU5URVJOQUxTX0RPX05PVF9VU0VfT1JfWU9VX1dJTExfQkVfRklSRUQlM0ElMjAlN0IlNUNuJTIwJTIwJTIwJTIwUmVhY3RDdXJyZW50T3duZXIlMkMlNUNuJTIwJTIwJTIwJTIwJTJGJTJGJTIwVXNlZCUyMGJ5JTIwcmVuZGVyZXJzJTIwdG8lMjBhdm9pZCUyMGJ1bmRsaW5nJTIwb2JqZWN0LWFzc2lnbiUyMHR3aWNlJTIwaW4lMjBVTUQlMjBidW5kbGVzJTNBJTVDbiUyMCUyMCUyMCUyMGFzc2lnbiUyQyU1Q24lMjAlMjAlN0QlMkMlNUNuJTdEJTNCJTVDbiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwUmVhY3QlM0IlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

Ideally this could just simplify to a `module.exports = React` call I think.

4. Value superposition inlining - instead of tracking known or unknown values, tracking values as a superposition of known values or value ranges (like a typing system) - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHAlMjAlM0QlMjB1bmtub3duR2xvYmFsJTIwJTNGJTIwJ2FzZGYnJTIwJTNBJTIwdHJ1ZSUzQiU1Q24lNUNuaWYlMjAoIXApJTVDbiU1Q3Rjb25zb2xlLmxvZygndHJlZXNoYWtlJTIwbWUnKSUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

5. Simple function inlining into its callsite (either based on single callsite, or some heuristic) - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwcSUyMChhJTJDJTIwYiklMjAlN0IlNUNuJTVDdHJldHVybiUyMGElMjAlMkIlMjBiJTNCJTVDbiU3RCU1Q25jb25zb2xlLmxvZyhxKCdhJyUyQyUyMCdiJykpJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

I know this is further off into the future, but worth thinking about as well.

6. Handling mutable variable timing - https://rollupjs.org/repl?version=0.58.1&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmxldCUyMG11dGFibGUlMjAlM0QlMjBmYWxzZSUzQiU1Q24lNUNuZnVuY3Rpb24lMjBwJTIwKCklMjAlN0IlNUNuJTVDdGlmJTIwKCFtdXRhYmxlKSU1Q24lNUN0JTVDdGNvbnNvbGUubG9nKCd0cmVlc2hha2UlMjBtZScpJTNCJTVDbiU3RCU1Q24lNUNubXV0YWJsZSUyMCUzRCUyMHRydWUlM0IlNUNucCgpJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmNqcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

One useful piece of analysis here might be to track execution order of each Node in the critical path execution so that assignments and calls on this critical path can be related on this timeline against mutable variable values for better optimization than just SSA cases.

The hard thing about tracking timings like this may be in dealing with gaps and ensuring we maintain knowledge of what could spill during gaps... seems like time complexity could grow there considerably unless we do something clever and shared with taint analysis tracking?

**Comment by guybedford** [comment]
Related here - what useful architectural refactorings might help extend our reach here? I guess its a constant process but if anything comes to mind would be interesting to hear. Some ideas:

* Further performance optimization and refactoring of existing treeshaking code, including say further analysis of hidden class or polymorphic deoptimizations
* Separate module format tests from analysis and treeshaking tests on a single module format output

**Comment by guybedford** [comment]
Update - -I've added another item to the list for value superpositions.

**Comment by lukastaegert** [comment]
For once, I would actually not call this "value inlining" but "node inlining" because that is what it would be in the end, and we should also try to preserve source maps while doing it.

That being said as I do not have a lot of time right now, here some things from the top of my backlog and what needs to be done. Note that I would be very happy if we do not do any critical path analysis just yet before everything else is in place as this will introduce quite a new level of complexity while providing only dubious returns. There is a TON of useful things that can be done without that.

Let's work backwards.

## 3. Node inlining
I would propose to inline each node that is
* only used once and which
* only depends on other nodes that can be inlined

Why the second condition? Because inlining can mean tearing variables out of their natural scopes so we need to make sure they either do not depend on their scope or can take their dependencies with them.

As for inlining nested object literal nodes, this can only happen if
* There have not been any unresolvable computed property accesses to one of the parent nodes
* The object has not been assigned to another object; this is because in the other object, the property would just be marked as UNKNOWN so we cannot track reassignments here
* The object has not been passed to a function because assignment tracking does not follow function parameters
* The object is not exported

I would propose to do the inlining (or "marking for inlining")  in a separate step between tree-shaking and rendering. That way, reference counting can take the tree-shaking results into account. A node that is actually used three times can be used only once in the active code and thus be eligible for inlining.

However to actually be successful, the code should be capable of removing parts of object literals so before doing the inlining, I would propose to do the following first:

## 2. Object literal tree-shaking
Currently, objects are included as a whole. However, we could when including a node of an object just include the specific node without any children and include children only in the following situations:
* The property is accessed
* A computed property is accessed
* The object has been assigned to another object (see above)
* The object has been passed to a function (see above)
* The object has been exported (see above)
In the last three cases, all children of children need to be included as well

However there is another situation that can be problematic: Object destructuring, as this can reach into objects in a new way. So first we should add

## 1. Tree-shaking for destructuring assignments
My best idea here is to add a new method to objects: `getNodeAtPath` (or: `getAssignmentAtPath`) which can be used to retrieve the initial value for destructured variable definitions out of their right hand assignment.

However since values can change once they are reassigned and we cannot guarantee a specific order in which reassignments are tracked, I would actually first

## 0. Bind and re-bind assignments during tree-shaking, not during initialization

If all these things are done in order, many of the more subtle propositions posed in the issue will be solved automatically but I do not have time to go into more detail.

Totally untouched of this, the following could be implemented at any time:

## -1. Generalize `getValueAtPath` truthiness tracking for conditionals

**Comment by guybedford** [comment]
> That being said as I do not have a lot of time right now, here some things from the top of my backlog and what needs to be done. Note that I would be very happy if we do not do any critical path analysis just yet before everything else is in place as this will introduce quite a new level of complexity while providing only dubious returns. There is a TON of useful things that can be done without that.

The goal of this discussion is to cast the net out as wide as we can first to envision the sort of architecture possibilities, and then come back to the immediate code scenarios so that we have some guiding architectural principles to keep in mind. There's no time pressure here at all. This space is huge as well, and we're going to be missing out large swathes of optimizations which will become clearer in future. But the core architecture to focus on is the combination of analysis and inlining passes in a way that can be flexible and scalable. This is such an important aspect of Rollup to ensure it is carefully designed to enable these possibilities.

I'd hope we can at least consider how we could enable execution analysis, even if it is a far-future item on the roadmap. Otherwise it seems like a tool that did perform this critical execution analysis could provide a superior output.

I'm happy with calling this Node inlining as well. If we move to a multi-pass inlining process I wonder if the time has come to handle each inlining along with its associated AST mutation to support chaining more complex combinations of inlings?

> There have not been any unresolvable computed property accesses to one of the parent nodes
> The object has not been assigned to another object; this is because in the other object, the property would just be marked as UNKNOWN so we cannot track reassignments here
> The object has not been passed to a function because assignment tracking does not follow function parameters
> The object is not exported

These rules effectively define a generic value-tracking taint analysis. Similar rules extend to determining if all the variables of a pure function are known in order to check that a function can truly be marked as pure from a Rollup perspective. Would definitely be a very useful analysis to have.

In terms of passes, the more passes we add the more things will slow down. As much as possible if we can determine the hard analysis like this during scope analysis itself in a single "scope and analysis" pass, that will likely help performance than running iterations each time a function is called. Perhaps the analysis caches information relating to determining the above on the nodes themselves, while the actual functions in the second pass then check the combination of these analysis indicators they are interested in. For example, setting `ObjectExpressionInstance.computedPropertyChildAccess = true / false` during the analysis pass (ideally part of a single scope/binding pass). Where individual member expressions can then check this root value without much perf cost at all.

Agreed it sounds like treeshaking / inlining will be separate passes unto themselves though. I wonder if inlining would enable new treeshaking to have their passes interleaved perhaps? eg `export default valueKnownToBeTrue || sideEffect()`, although will have to see how useful these patterns are.

> Currently, objects are included as a whole. However, we could when including a node of an object just include the specific node without any children and include children only in the following situations:
> The property is accessed
> A computed property is accessed
> The object has been assigned to another object (see above)
> The object has been passed to a function (see above)
> The object has been exported (see above)
> In the last three cases, all children of children need to be included as well

It seems like you're adding the exceptions here which were exactly the taint analysis indicators, but I'm not sure I see any benefit to inlining an object along with its children though? What use cases do you see for this? I am personally happy to focus on property inlining when it can work.

There are a whole bunch of considerations around object inlining, for example if a single object property is used in multiple places, we could treat that as an extraction or object tree-shaking over an inlining:

Eg: ```js
var obj = { a: 'a', b: 'b' };

console.log(obj.b);
console.log(obj.b);
```

We could, for example treat it as an extraction of `const b = b`, with all usage of `obj.b` being replaced with `var obj = { a: 'a', b }`. If we did multi-pass with AST mutation, then removing the object would become a natural part of a subsequent tree-shaking pass as opposed to having all the "object inlining" logic in one place.

Probably worth opening a separate issue to discuss these sorts of cases in more detail, but let's keep working backwards from far-future goals for now.

> Tree-shaking for destructuring assignments

Definitely sounds like a worthwhile next step to me.

Rather than a get function I wonder if it would be nice to move towards this information being "populated" during the scope and analysis pass.

For example in the code `var { a: { b } } = { a: { b: 5 } }`, the Variable `b` would be defined as having an assignment expression of the literal `5` (eg something like `VariableInstance.assignments = [{ expr: LiteralNode, default: undefined }]`, similar to the analysis done by EScope - http://mazurov.github.io/escope-demo/). We can then handle inlining through reading this information off the variable itself, without having to possibly traverse the destructuring pattern multiple times.

> Bind and re-bind assignments during tree-shaking, not during initialization

Please do clarify on this when you can. No rush to respond - I will likely only have time to look into working on features here myself in a couple of weeks too.

**Comment by guybedford** [comment]
I thought it would be worth following on from https://github.com/rollup/rollup/pull/2167#discussion_r186616711 here since this issue is covering future optimization opportunities.

In short, the transform I am using to replicate CommonJS execution semantics looks something like the following:

```js
// exports.exportName = 'value';
var moduleAExports = {};
var isModuleAExecuted = false;
function executeModuleA () {
  isModuleAExecuted = true;
  moduleAExports.exportName = 'value';
}

// require('a');
var moduleBExports = {};
var isModuleBExecuted = false;
function executeModuleB () {
  isModuleBExecuted = true;
  (!isModuleAExecuted && executeModuleA() || moduleAExports) // require('a')
}

// begin top-level instantiate
if (!isModuleBExecuted)
  executeModuleB();
```

Note how a lot of the flags above can be simplified if we know the whole tree execution behaviour, but from the perspective of a single module plugin transform, the full transform is needed to get proper support.

@lukastaegert would really value your input on whether Rollup might support optimization of the above, or if there is a way to rewrite the above to better support what optimizations might be possible in future.

**Comment by guybedford** [comment]
Prepack gets this amazingly - https://prepack.io/repl.html#BQMwrgdgxgLglgewgAmASmQbwFDYPR7ICmAHgA4IBOMAzgHSkXUByAhgLZHIC8yA5ADdWAGzBE+AbmxDKydggAmYYUQCCAUXJVaPLAF8pM5HBoBZRcrWaiUMDCILdIETSJTw0eEmIkbdouZKKqqoGDjIxmYWwda29o68MJRiUhHyQVZa1PSM2mycuoIiYpLYergEyFBINAgqdMIIAObAlEQAjmBwbcB8rHxoaIassumWAEKaTDq8mAbSI5GBE7H+CcjOwq7ukLCIKKR+9ssq46FY2BEmJ0STvnEOukkpl1U1dUQNzcAAhNfRmSOjwAZMCfECbqp0MgAD4wuQAjRZWhoMoVQgAIyITTgKBgCDIAFoVAIiMJjBAaDBWBB4Kx7Ng4CBUH8ohk7kCFKiIocHjdxuhUsgymhBUA

Rollup + Prepack seems like the preferred solution here, but the issue then is that we miss out on these optimizations in code splitting workflows without whole-graph analysis.

**Comment by zenmumbler** [comment]
I was going to post this at #2167, but this seems to be a continuation of the discussion.

This is amazing work and starting to look like it can handle some functionality often done by a preprocessor. I'm assuming this is not covered by this PR but wanted to check. What would happen to something like this:

```js
const DEBUG_MODE = false; // set by an env var or something
function assert(condition, message) {
  if (DEBUG_MODE) {
    if (! condition) {
      throw new Error(message);
    }
  }
}

// elsewhere:
function x(value) {
  assert(value >= 0, "value must be positive");
  console.log("VAL", value);
}

const q = parseInt(prompt("Gimme a number!"));
x(q);
```

In this case the whole body of assert would be omitted, yielding an empty function. Ideally assert itself and all of its invocations would then also be omitted as they are no-ops.

I think this is maybe a bridge too far but this would really make for some powerful build-time switches to have easy and clean debug and production builds among other things.  Prepack reduces the code to a single console.log but complains that it may throw, which is kind of the whole point.

Anyway, very exciting stuff!

**Comment by zenmumbler** [comment]
@lukastaegert Excellent! I am going to be happy with this outcome and not immediately muse about all kinds of _other_ compiler optimisations and stuff these kinds of transformations could handle... 😇

**Comment by guybedford** [comment]
I think this discussion has reached its conclusions in considering the tree-shaking roadmap for now.

I've split this out into the separate issues at https://github.com/rollup/rollup/issues/2201, https://github.com/rollup/rollup/issues/2202,https://github.com/rollup/rollup/issues/2203. Just let me know if I've missed anything here too.
